test/src/asyncStressTest.cpp

```diff
@@ -0,0 +1,242 @@
+			transformNodes.push_back(childNode);
+		}
+
+		setRandomRays();
+	}
+
+	void setRandomRays()
+	{
+		std::uniform_real_distribution<float> randomFloat {-0.9, 0.9};
+		std::uniform_int_distribution<size_t> randomInt {1, 10};
+		raysXY = { randomFloat(randomGenerator), randomFloat(randomGenerator) };
+		rayCount = { randomInt(randomGenerator), randomInt(randomGenerator) };
+		std::vector<rgl_mat3x4f> rays = makeGridOfParallelRays(raysXY, raysXY, {rayCount.x(), rayCount.y()});
+		EXPECT_RGL_SUCCESS(rgl_node_rays_from_mat3x4f(&useRaysNode, rays.data(), rays.size()));
+	}
+
+	std::unordered_map<rgl_node_t, Mat3x4f> getCumulativeTransform() const
+	{
+		std::unordered_map<rgl_node_t, Mat3x4f> result;
+		std::function<void(rgl_node_t, Mat3x4f)> dfs = [&](rgl_node_t currentBase, Mat3x4f parentCumulative) {
+			auto* currentTyped = dynamic_cast<TransformPointsNode*>(currentBase);
+			Mat3x4f currentCumulative = currentTyped->getTransform() * parentCumulative;
+			result.insert({currentBase, currentCumulative});
+			for(auto&& childBase : currentBase->getOutputs()) {
+				if (!result.contains(childBase.get())) {
+					dfs(childBase.get(), currentCumulative);
+				}
+			}
+		};
+		dfs(parentTransformNode, Mat3x4f::identity());
+		return result;
+	}
+};
+
+/* Helper structure for data associated with a single graph and a single run-modify-check cycle */
+struct GraphRun {
+	int expectedPointCount;
+	Vec3f expectedPointWorld;
+	rgl_node_t updateRaysAfterCheckingNode;
+	std::unordered_map<rgl_node_t, Mat3x4f> expectedTransform;
+	std::vector<int> nodeCheckOrder;
+	int nextNodeCheckOrderIndex;
+};
+
+/**
+ * The test runs a graph and concurrently to its execution, it
+ * - changes numbers of rays of UseRaysNode
+ * - queries output from a subset of random TransformPointsNodes
+ * - may call rgl_graph_run while it is still running
+ * The test aims to detect any inconsistency that would be caused by race condition bug.
+ *
+ * TODO: Possible improvements:
+ * - When the graph is running, do also scene modifications to stress scene / graph concurrency.
+ *
+ * - Except modifying just the first node, modify also parameters (transforms) of TransformPointsNodes
+ *   - May deteriorate testing concurrency, because requires a relatively long update of cumulative transform map.
+ */
+TEST_F(GraphStress, Async)
+{
+	using namespace testing;
+
+	// Config
+	int GRAPH_COUNT = 16;  // Number of graphs executed in parallel
+	const int GRAPH_SIZE = 1024;  // Number of TransformPointsNodes in each graph
+	const int GRAPH_QUERIES = 32;  // How many nodes are checked for results (should be a subset)
+	const int GRAPH_EPOCHS = 32;  // How many times all graphs are executed, changed and queried
+	const float EPSILON = 2 * 1E-6;  // Tolerance for GPU-CPU matrix multiplication differences
+	ASSERT_THAT(GRAPH_QUERIES, Lt(GRAPH_SIZE));
+
+	// Optional logging
+	// EXPECT_RGL_SUCCESS(rgl_configure_logging(RGL_LOG_LEVEL_ALL, nullptr, true));
```

Comment:
I would prefer some local DEBUG_WITH_LOGS macro in the file, instead of commented code.

Reply:
This will be addressed by @nebraszka. She already did something like this in her PR.

Reply:
Jk jk, we remove logging from tests.